src/Adapter/Product/Shop/CommandHandler/SetProductShopsHandler.php

```diff
@@ -0,0 +1,157 @@
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Product\Shop\CommandHandler;
+
+use PrestaShop\PrestaShop\Adapter\Product\ProductDeleter;
+use PrestaShop\PrestaShop\Adapter\Product\Repository\ProductRepository;
+use PrestaShop\PrestaShop\Adapter\Product\Update\ProductShopUpdater;
+use PrestaShop\PrestaShop\Adapter\Shop\Repository\ShopRepository;
+use PrestaShop\PrestaShop\Core\Domain\Product\Command\DeleteProductCommand;
+use PrestaShop\PrestaShop\Core\Domain\Product\Exception\InvalidProductShopAssociationException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Shop\Command\SetProductShopsCommand;
+use PrestaShop\PrestaShop\Core\Domain\Product\Shop\CommandHandler\SetProductShopsHandlerInterface;
+use PrestaShop\PrestaShop\Core\Domain\Shop\Exception\ShopException;
+use PrestaShop\PrestaShop\Core\Domain\Shop\ValueObject\ShopId;
+
+class SetProductShopsHandler implements SetProductShopsHandlerInterface
+{
+    /**
+     * @var ShopRepository
+     */
+    private $shopRepository;
+
+    /**
+     * @var ProductRepository
+     */
+    private $productRepository;
+
+    /**
+     * @var ProductDeleter
+     */
+    private $productDeleter;
+
+    /**
+     * @var ProductShopUpdater
+     */
+    private $productShopUpdater;
+
+    public function __construct(
+        ShopRepository $shopRepository,
+        ProductRepository $productRepository,
+        ProductDeleter $productDeleter,
+        ProductShopUpdater $productShopUpdater
+    ) {
+        $this->shopRepository = $shopRepository;
+        $this->productRepository = $productRepository;
+        $this->productDeleter = $productDeleter;
+        $this->productShopUpdater = $productShopUpdater;
+    }
+
+    public function handle(SetProductShopsCommand $command): void
+    {
+        $productId = $command->getProductId();
+        $sourceShopId = $command->getSourceShopId();
+        $selectedShopIds = $command->getShopIds();
+
+        if (empty($selectedShopIds)) {
+            throw new InvalidProductShopAssociationException(
+                sprintf(
+                    'Empty shop association provided. Use %s command to delete product instead',
+                    DeleteProductCommand::class
+                )
+            );
+        }
```

Comment:
This can be handled directly in the command constructor, no need to wait for the handler to detect this. And you can either use this new exception or an `InvalidArgumentException` as you prefer

Reply:
:+1: I'll leave this new exception as it helps me to distinguish this error in behat tests